scylla/src/transport/load_balancing/default.rs

```diff
@@ -272,13 +321,23 @@ impl DefaultPolicy {
             .filter(predicate)
     }
 
+    #[allow(clippy::type_complexity)]
     fn pick_replica<'a>(
         &'a self,
         ts: &TokenWithStrategy<'a>,
-        should_be_local: bool,
+        replica_location: ReplicaLocation,
         predicate: &impl Fn(&NodeRef<'a>) -> bool,
         cluster: &'a ClusterData,
     ) -> Option<NodeRef<'a>> {
+        let (predicate, should_be_local): (Box<dyn Fn(&NodeRef<'a>) -> bool>, bool) =
+            match replica_location {
+                ReplicaLocation::Any => (Box::new(|node| predicate(node)), false),
+                ReplicaLocation::Datacenter => (Box::new(|node| predicate(node)), true),
+                ReplicaLocation::DatacenterAndRack => (
+                    Box::new(|node| predicate(node) && node.rack == self.preferred_rack),
+                    true,
+                ),
+            };
```

Comment:
As the "boxing problem" is a possible inefficiency, this is my proposed solution:
(instead of having three closures with one behaviour specified, let's have a single predicate closure that chooses its behaviour with a `match`)
```rust
  fn pick_replica<'a>(
      &'a self,
      ts: &TokenWithStrategy<'a>,
      replica_location: ReplicaLocation,
      predicate: &impl Fn(&NodeRef<'a>) -> bool,
      cluster: &'a ClusterData,
  ) -> Option<NodeRef<'a>> {
      let predicate = |node| match replica_location {
          ReplicaLocation::Any |
          ReplicaLocation::Datacenter => predicate(&node),
          ReplicaLocation::DatacenterAndRack => predicate(&node) && node.rack == self.preferred_rack,
      };
      let should_be_local = match replica_location {
          ReplicaLocation::Any => false,
          ReplicaLocation::Datacenter |
          ReplicaLocation::DatacenterAndRack => true,
      };
      let replica_set = self.nonfiltered_replica_set(ts, should_be_local, cluster);
      if let Some(fixed) = self.fixed_shuffle_seed {
          let mut gen = Pcg32::new(fixed, 0);
          replica_set.choose_filtered(&mut gen, predicate)
      } else {
          replica_set.choose_filtered(&mut thread_rng(), predicate)
      }
  }
```

Reply:
Done. Thanks for the hint.